src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -126,96 +125,138 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
                     _reindexJobRecord.Count = countOnlyResults.TotalCount.Value;
 
                     // Query first batch of resources
-                    await ProcessQueryAsync(queryStatus, cancellationToken);
+                    await ProcessQueryAsync(queryStatus, jobLock, cancellationToken);
                 }
-                else
+
+                var queryTasks = new ConcurrentBag<Task>();
+
+                // while not all queries are finished
+                while (_reindexJobRecord.QueryList.Where(q =>
+                    q.Status == OperationStatus.Queued ||
+                    q.Status == OperationStatus.Running).Any())
                 {
-                    // check to see if queries are queued
-                    // TODO: this while loop is temporary until we multithread this task so multiple threads can
-                    // processes queries
-                    while (_reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).Any())
+                    if (_reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).Any())
                     {
                         // grab the next query from the list which is labeled as queued and run it
                         var query = _reindexJobRecord.QueryList.Where(q => q.Status == OperationStatus.Queued).OrderBy(q => q.LastModified).FirstOrDefault();
 
-                        await ProcessQueryAsync(query, cancellationToken);
+#pragma warning disable CS4014 // Suppressed as we want to continue execution and begin processing the next query while this continues to run
+                        queryTasks.Add(ProcessQueryAsync(query, jobLock, cancellationToken));
+                        _logger.LogInformation($"Reindex job task using {queryTasks.Count} threads");
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+                    }
+
+                    // reset stale queries to pending
+                    var staleQueries = _reindexJobRecord.QueryList.Where(
+                        q => q.Status == OperationStatus.Running && q.LastModified < DateTimeOffset.UtcNow - _reindexJobConfiguration.JobHeartbeatTimeoutThreshold);
+                    foreach (var staleQuery in staleQueries)
+                    {
+                        lock (jobLock)
+                        {
+                            staleQuery.Status = OperationStatus.Queued;
+                            UpdateJobAsync(cancellationToken).Wait();
+                        }
+                    }
+
+                    await Task.Delay(_reindexJobConfiguration.QueryDelayIntervalInMilliseconds);
+
+                    if (queryTasks.Count >= reindexJobRecord.MaximumConcurrency)
+                    {
+                        var taskArray = queryTasks.ToArray();
+                        int taskIndex = Task.WaitAny(taskArray);
```

Comment:
Let's say `MaximumConcurrency = 5`. Even if all tasks are complete, we will only remove one task. That means the next loop can add only one task. Which means after this initial condition `queryTasks.Count >= reindexJobRecord.MaximumConcurrency` is met, we will only be adding/working on one task at a time.

Reply:
updated to remove multiple tasks.